File: Tree/BinaryTree/Traversal.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

class TreeNode {
public:
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int data) {
        this->data = data;
        left = nullptr;
        right = nullptr;
    }
};

/*
Function: insertion()
Purpose : Builds the binary tree recursively using user input (-1 = no node)

Time Complexity:
    O(N) — Each node is created exactly once.
Space Complexity:
    O(H) — Recursive stack space (H = height of tree)
*/
TreeNode* insertion() {
    int val;
    cout << "Enter the data (-1 for no node): ";
    cin >> val;
    if (val == -1)
        return nullptr;

    TreeNode* node = new TreeNode(val);
    cout << "Enter the left child of " << val << ": ";
    node->left = insertion();

    cout << "Enter the right child of " << val << ": ";
    node->right = insertion();

    return node;
}

/*
Function: inorder()
Purpose : Recursive Inorder Traversal (Left, Root, Right)

Time Complexity: O(N)
Space Complexity: O(H)
*/
void inorder(TreeNode* root) {
    if (root == nullptr)
        return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
}

/*
Function: preorder()
Purpose : Recursive Preorder Traversal (Root, Left, Right)

Time Complexity: O(N)
Space Complexity: O(H)
*/
void preorder(TreeNode* root) {
    if (root == nullptr)
        return;
    cout << root->data << " ";
    preorder(root->left);
    preorder(root->right);
}

/*
Function: postorder()
Purpose : Recursive Postorder Traversal (Left, Right, Root)

Time Complexity: O(N)
Space Complexity: O(H)
*/
void postorder(TreeNode* root) {
    if (root == nullptr)
        return;
    postorder(root->left);
    postorder(root->right);
    cout << root->data << " ";
}

/*
Function: levelorder()
Purpose : BFS traversal of all levels using a queue

Time Complexity: O(N)
Space Complexity: O(N)
*/
vector<vector<int>> levelorder(TreeNode* root) {
    vector<vector<int>> ans;
    if (root == nullptr) return ans;

    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int size = q.size();
        vector<int> level;
        for (int i = 0; i < size; i++) {
            TreeNode* node = q.front();
            q.pop();
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
            level.push_back(node->data);
        }
        ans.push_back(level);
    }
    return ans;
}

/*
Function: preorderitr()
Purpose : Iterative Preorder Traversal using a stack

Time Complexity: O(N)
Space Complexity: O(H) — stack stores nodes (H = height of tree)
*/
vector<int> preorderitr(TreeNode* root) {
    vector<int> preorder;
    if (root == nullptr) return preorder;

    stack<TreeNode*> st;
    st.push(root);
    while (!st.empty()) {
        TreeNode* node = st.top();
        st.pop();
        preorder.push_back(node->data);

        // Push right child first so left is processed first
        if (node->right != nullptr) st.push(node->right);
        if (node->left != nullptr) st.push(node->left);
    }
    return preorder;
}

/*
Function: inorderitr()
Purpose : Iterative Inorder Traversal using a stack

Time Complexity: O(N)
Space Complexity: O(H)
*/
vector<int> inorderitr(TreeNode* root) {
    vector<int> inorder;
    stack<TreeNode*> st;
    TreeNode* node = root;

    while (true) {
        if (node != nullptr) {
            st.push(node);
            node = node->left;
        } else {
            if (st.empty()) break;
            node = st.top();
            st.pop();
            inorder.push_back(node->data);
            node = node->right;
        }
    }
    return inorder;
}

/*
Function: postorderitr()
Purpose : Iterative Postorder Traversal using two stacks

Time Complexity: O(N)
Space Complexity: O(2H) ≈ O(H)
*/
vector<int> postorderitr(TreeNode* root) {
    vector<int> postorder;
    if (root == nullptr) return postorder;

    stack<TreeNode*> st1, st2;
    st1.push(root);

    while (!st1.empty()) {
        TreeNode* node = st1.top();
        st1.pop();
        st2.push(node);

        if (node->left) st1.push(node->left);
        if (node->right) st1.push(node->right);
    }

    while (!st2.empty()) {
        postorder.push_back(st2.top()->data);
        st2.pop();
    }
    return postorder;
}
/*
Function: postorderitr_1stack()
Purpose : Iterative Postorder Traversal using only one stack.

Logic:
- Traverse left as much as possible.
- When no left child exists:
    - Check right child of the top node.
    - If no right child → process the node (add to postorder).
    - If right child exists → move to right subtree.
- Use `temp` to track recently visited node so we don't revisit right subtrees.

Time Complexity: O(N)
Space Complexity: O(H)
*/
vector<int> postorderitr_1stack(TreeNode* root) {
    vector<int> post;
    if (root == nullptr) return post;

    stack<TreeNode*> st;
    TreeNode* curr = root;
    TreeNode* temp = nullptr;

    while (curr != nullptr || !st.empty()) {
        if (curr != nullptr) {
            st.push(curr);
            curr = curr->left;
        } else {
            temp = st.top()->right;
            if (temp == nullptr) {
                // No right child → process node
                temp = st.top();
                st.pop();
                post.push_back(temp->data);

                // Keep popping while current node is right child of top
                while (!st.empty() && temp == st.top()->right) {
                    temp = st.top();
                    st.pop();
                    post.push_back(temp->data);
                }
            } else {
                // Move to right child
                curr = temp;
            }
        }
    }
    return post;
}


int main() {
    TreeNode* root = nullptr;
    root = insertion();

    cout << "\nRecursive Traversals:" << endl;
    cout << "Inorder: ";
    inorder(root);
    cout << "\nPreorder: ";
    preorder(root);
    cout << "\nPostorder: ";
    postorder(root);

    cout << "\n\nIterative Traversals:" << endl;

    vector<int> pre = preorderitr(root);
    cout << "Preorder (Iterative): ";
    for (int val : pre) cout << val << " ";
    cout << endl;

    vector<int> ino = inorderitr(root);
    cout << "Inorder (Iterative): ";
    for (int val : ino) cout << val << " ";
    cout << endl;

    vector<int> post = postorderitr(root);
    cout << "Postorder (Iterative): ";
    for (int val : post) cout << val << " ";
    cout << endl;

    cout << "\nLevel Order Traversal:\n";
    vector<vector<int>> levels = levelorder(root);
    for (auto& level : levels) {
        for (int val : level)
            cout << val << " ";
        cout << endl;
    }

    cout << endl;
    return 0;
}

File: java/Tree/BinaryTree/Traversal.java
package Tree.BinaryTree;

import java.util.*;

/**
 * Tree Node Definition
 */
class TreeNode {
    int data;
    TreeNode left;
    TreeNode right;

    TreeNode(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

public class Traversal{

    static Scanner sc = new Scanner(System.in);

    /*
     * Function: insertion()
     * Purpose : Builds the binary tree recursively using user input (-1 = no node)
     *
     * Time Complexity:
     *     O(N) — Each node is created exactly once.
     * Space Complexity:
     *     O(H) — Recursive stack space (H = height of tree)
     */
    static TreeNode insertion() {
        System.out.print("Enter the data (-1 for no node): ");
        int val = sc.nextInt();

        if (val == -1)
            return null;

        TreeNode node = new TreeNode(val);

        System.out.println("Enter the left child of " + val + ": ");
        node.left = insertion();

        System.out.println("Enter the right child of " + val + ": ");
        node.right = insertion();

        return node;
    }

    /*
     * Function: inorder()
     * Purpose : Recursive Inorder Traversal (Left, Root, Right)
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static void inorder(TreeNode root) {
        if (root == null)
            return;
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }

    /*
     * Function: preorder()
     * Purpose : Recursive Preorder Traversal (Root, Left, Right)
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static void preorder(TreeNode root) {
        if (root == null)
            return;
        System.out.print(root.data + " ");
        preorder(root.left);
        preorder(root.right);
    }

    /*
     * Function: postorder()
     * Purpose : Recursive Postorder Traversal (Left, Right, Root)
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static void postorder(TreeNode root) {
        if (root == null)
            return;
        postorder(root.left);
        postorder(root.right);
        System.out.print(root.data + " ");
    }

    /*
     * Function: levelorder()
     * Purpose : BFS traversal of all levels using a queue
     *
     * Time Complexity: O(N)
     * Space Complexity: O(N)
     */
    static List<List<Integer>> levelorder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null)
            return ans;

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            int size = q.size();
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                level.add(node.data);

                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }
            ans.add(level);
        }
        return ans;
    }

    /*
     * Function: preorderitr()
     * Purpose : Iterative Preorder Traversal using a stack
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static List<Integer> preorderitr(TreeNode root) {
        List<Integer> preorder = new ArrayList<>();
        if (root == null)
            return preorder;

        Stack<TreeNode> st = new Stack<>();
        st.push(root);

        while (!st.isEmpty()) {
            TreeNode node = st.pop();
            preorder.add(node.data);

            // Push right first so left is processed first
            if (node.right != null)
                st.push(node.right);
            if (node.left != null)
                st.push(node.left);
        }
        return preorder;
    }

    /*
     * Function: inorderitr()
     * Purpose : Iterative Inorder Traversal using a stack
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static List<Integer> inorderitr(TreeNode root) {
        List<Integer> inorder = new ArrayList<>();
        Stack<TreeNode> st = new Stack<>();
        TreeNode node = root;

        while (true) {
            if (node != null) {
                st.push(node);
                node = node.left;
            } else {
                if (st.isEmpty())
                    break;
                node = st.pop();
                inorder.add(node.data);
                node = node.right;
            }
        }
        return inorder;
    }

    /*
     * Function: postorderitr()
     * Purpose : Iterative Postorder Traversal using two stacks
     *
     * Time Complexity: O(N)
     * Space Complexity: O(2H) ≈ O(H)
     */
    static List<Integer> postorderitr(TreeNode root) {
        List<Integer> postorder = new ArrayList<>();
        if (root == null)
            return postorder;

        Stack<TreeNode> st1 = new Stack<>();
        Stack<TreeNode> st2 = new Stack<>();

        st1.push(root);

        while (!st1.isEmpty()) {
            TreeNode node = st1.pop();
            st2.push(node);

            if (node.left != null)
                st1.push(node.left);
            if (node.right != null)
                st1.push(node.right);
        }

        while (!st2.isEmpty()) {
            postorder.add(st2.pop().data);
        }
        return postorder;
    }

    /*
     * Function: postorderitr_1stack()
     * Purpose : Iterative Postorder Traversal using only one stack
     *
     * Time Complexity: O(N)
     * Space Complexity: O(H)
     */
    static List<Integer> postorderitr_1stack(TreeNode root) {
        List<Integer> post = new ArrayList<>();
        if (root == null)
            return post;

        Stack<TreeNode> st = new Stack<>();
        TreeNode curr = root;
        TreeNode temp;

        while (curr != null || !st.isEmpty()) {
            if (curr != null) {
                st.push(curr);
                curr = curr.left;
            } else {
                temp = st.peek().right;
                if (temp == null) {
                    temp = st.pop();
                    post.add(temp.data);

                    while (!st.isEmpty() && temp == st.peek().right) {
                        temp = st.pop();
                        post.add(temp.data);
                    }
                } else {
                    curr = temp;
                }
            }
        }
        return post;
    }

    public static void main(String[] args) {

        TreeNode root = insertion();

        System.out.println("\nRecursive Traversals:");
        System.out.print("Inorder: ");
        inorder(root);

        System.out.print("\nPreorder: ");
        preorder(root);

        System.out.print("\nPostorder: ");
        postorder(root);

        System.out.println("\n\nIterative Traversals:");

        System.out.print("Preorder (Iterative): ");
        for (int val : preorderitr(root))
            System.out.print(val + " ");
        System.out.println();

        System.out.print("Inorder (Iterative): ");
        for (int val : inorderitr(root))
            System.out.print(val + " ");
        System.out.println();

        System.out.print("Postorder (Iterative): ");
        for (int val : postorderitr(root))
            System.out.print(val + " ");
        System.out.println();

        System.out.println("\nLevel Order Traversal:");
        for (List<Integer> level : levelorder(root)) {
            for (int val : level)
                System.out.print(val + " ");
            System.out.println();
        }
    }
}
